flang/test/Lower/OpenMP/derived-type-map.f90

```diff
@@ -0,0 +1,105 @@
+!CHECK: %[[DECLARE:.*]]:2 = hlfir.declare %[[ALLOCA]] {uniq_name = "_QFmaptype_derived_explicitEscalar_arr"} : (!fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>) -> (!fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>, !fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>)
+!CHECK: %[[MAP:.*]] = omp.map_info var_ptr(%[[DECLARE]]#0 : !fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>, !fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>) map_clauses(tofrom) capture(ByRef) -> !fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>> {name = "scalar_arr"}
+!CHECK:  omp.target map_entries(%[[MAP]] -> %[[ARG0:.*]] : !fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>) {
+!CHECK:    ^bb0(%[[ARG0]]: !fir.ref<!fir.type<_QFmaptype_derived_explicitTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>):
+subroutine mapType_derived_explicit
+    type :: scalar_and_array
+      real(4) :: real
+      integer(4) :: array(10)
+      integer(4) :: int
+    end type scalar_and_array
+    type(scalar_and_array) :: scalar_arr 
+    
+    !$omp target map(tofrom: scalar_arr)
+       scalar_arr%int = 1
+    !$omp end target
+end subroutine mapType_derived_explicit
+
+!CHECK: %[[ALLOCA:.*]] = fir.alloca !fir.type<_QFmaptype_derived_explicit_single_memberTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}> {bindc_name = "scalar_arr", uniq_name = "_QFmaptype_derived_explicit_single_memberEscalar_arr"}
+!CHECK: %[[DECLARE:.*]]:2 = hlfir.declare %[[ALLOCA]] {uniq_name = "_QFmaptype_derived_explicit_single_memberEscalar_arr"} : (!fir.ref<!fir.type<_QFmaptype_derived_explicit_single_memberTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>) -> (!fir.ref<!fir.type<_QFmaptype_derived_explicit_single_memberTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>, !fir.ref<!fir.type<_QFmaptype_derived_explicit_single_memberTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>)
+!CHECK: %[[MEMBER:.*]] = hlfir.designate %[[DECLARE]]#0{"array"}   shape %{{.*}} : (!fir.ref<!fir.type<_QFmaptype_derived_explicit_single_memberTscalar_and_array{real:f32,array:!fir.array<10xi32>,int:i32}>>, !fir.shape<1>) -> !fir.ref<!fir.array<10xi32>>
+!CHECK: %[[BOUNDS:.*]] = omp.bounds lower_bound(%{{.*}} : index) upper_bound(%{{.*}} : index) extent(%{{.*}} : index) stride(%{{.*}} : index) start_idx(%{{.*}} : index)
+!CHECK: %[[MEMBER_MAP:.*]] = omp.map_info var_ptr(%[[MEMBER]] : !fir.ref<!fir.array<10xi32>>, !fir.array<10xi32>) map_clauses(tofrom) capture(ByRef) bounds(%[[BOUNDS]]) -> !fir.ref<!fir.array<10xi32>> {name = "scalar_arr%array"}
```

Comment:
Hey @razvanlupusoru sorry to bother you, but I had a little question relating to the var_ptr/var_ptr_ptr assignment for the member mappings, currently I've opted to use var_ptr in this PR for the offload pointer, which I am aware is likely incorrect and I am more than happy to fix. However, in this case, if I utilise the base address returned by gatherDataOperandAddrAndBounds for the var_ptr_ptr field which I expect is more correct, what would be the appropriate value to store in the var_ptr field? My FIR/HLFIR knowledge is unfortunately rather lackluster, so I was a little unsure what the correct series of operations to get the desired var_ptr field would be. Again, I am sorry for the bother and thank you very much for your help ahead of time!

Reply:
`gatherDataOperandAddrAndBounds` is giving you the pointer the value you want to map so it's likely to be the one you want to used in var_ptr and not var_ptr_ptr

Reply:
Thank you very much @clementval! Then I believe the PR may be correct in this aspect in it's current form, if not then I am more than happy to adjust it. I apologies for the insistent questions about this and I very much appreciate the help you and everyone else provides.

Reply:
I understand your question and no need to apologies :-) I guess you want to plug this directly to what the OMPIRBuilder is expecting (2 pointers and a size IIRC).

I'll have a deeper look early next week so I can double check we are on the same page.

Reply:
There is no pointer attachment (`PTR_AND_OBJ`) involved, so `var_ptr_ptr` should be unused.  The pointer of `var_ptr` is the starting address of `array`, which is a part of the `scalar_arr` structured scalar.  So this `map_info` looks correct to me.

But, overall, I am not sure why we need to handle this particular member array differently from other members of the derived type: all the members are mapped as part of the structure same as in https://github.com/llvm/llvm-project/blob/3a3302ef7b48f7907d0fb62b380d9d515a5f35e4/clang/lib/CodeGen/CGOpenMPRuntime.cpp#L6879

It would make sense to have the member map info if pointer attachment was involved, e.g. if `array` was a member represented with a descriptor.

Reply:
Oh, I was looking at the wrong piece of source code (above).  I see what you are doing with the member map infos now.  Please ignore the second part of my comment.

Reply:
Looks right to me as-is :)

Reply:
Thank you everybody for the input, I greatly appreciate it. If you see anything odd on a later inspection after I've fragmented the PR into something more byte sized and easy to consume do please mention it!

> I understand your question and no need to apologies :-) I guess you want to plug this directly to what the OMPIRBuilder is expecting (2 pointers and a size IIRC).
>

Yes, that's exactly what I am trying to do! :-) most of the OpenMP map work (at least lately) has been primarily working out how to reasonably pass data down to the OMPIRBuilder through (Flang -> MLIR -> LLVMIR) to fill the kernel argument structure (offload pointer/base pointer/size/map type being the primary fields) as it expects for the various data types. Whilst trying to keep the dialect open to use with the possible future C/C++ dialect and now I suppose intermixing with the GPU offload dialect as there's some push towards that! 

> I'll have a deeper look early next week so I can double check we are on the same page.

No rush, take your time! And thank you very much for your help and time.